### ISSUE TITLE
REQUEST: Different merge policy for breaking API changes

### ISSUE THREAD

**Issue by stevenroose** [rust-bitcoin/rust-bitcoin#2482]
I'm tired of noticing unneeded breaking changes to the API. It's one of the main reasons why most of my projects are using some older version of rust-bitcoin. It's just a headache to update an entire stack.

But I don't have time to keep up with everything that happens here (since I last kept up in end of September I not have >6000 e-mail notifications to go through).

I do realize that it sucks if minor changes or pure additions etc have to be pending for ages, but breaking changes really suck. The "we're just working towards 1.0 and then it'll be stable" excuse doesn't fly for 2 years in a row. This library is worth too much for me, both as a dependency but also emotionally, that it hurts me that I've started hating a lot of the changes in the last year or two.

Can we have some kind of special treatment of breaking changes? Give them a special tag in the MR list and leave them open for at least a month?

(For those using e-mail notifications: I don't think you can filter on tags using e-mail, you could with RSS I think, but if the load isn't insanely huge like with all notifications, it would be OK to take a day a month to go over them inside GitHub.)

**Comment by apoelstra** [comment]
We're just working towards 1.0 and then it will be stable.

**Comment by stevenroose** [comment]
> it was in a very bad shape when I came

I think this is the root disagreement here. I had used rust-bitcoin extensively for over 4 years before you joined and it did the job well. Sure the API wasn't super fancy and could use some love, but just using the official docs, you could do anything you would need to do. I don't know if you worked much on larger Rust-based Bitcoin projects, but *stability to me is way more important than getting the API 100% right*.

Instead of being "conforming Rust style guidelines" or "have a nitty-griddy API", **I think a library like this should have as main goal to have as many eyes on the code as possible**. And right now it is achieving the exact opposite. Review burden should be as low as possible, because the code in this project is just so critical to many people's projects. Whenever you think of a meaningful change (nits and internal optimizations excluded), we should consider whether the change is worth the review burden on the reviewers.

Currently there are 3 main active contributors (I believe), none of which I think are active users of the library in applications. I might be mistaken in this of course, not all our work is on public GitHub, but if true, that's not an ideal situation, especially not if you're actually burdening the users with breaking changes to supposedly make developer experience better.

**Comment by apoelstra** [comment]
FWIW I do maintain applications. Several things got stuck on 0.29 because we released 0.30 without any hex support, which I won't let happen again.

For the most part, releases are steps along the way to 1.0. You don't need to keep up with all of them in real time.

**Comment by apoelstra** [comment]
I just updated my trading bot from 0.29 to 0.31. Going to 0.30 was a bit of a PITA because I had to add a `hex-conservative` dep, and also the `Address<Unchecked>` stuff is a rough transition (but OTOH the new API now forces me to check networks, which previously I was not bothering to do at all, which is a nice sanity check to have).

Then 0.30 to 0.31 let me get rid of the `hex-conservative` dep in favor of `bitcoin::hex`. I still have a dep on the `hex` crate for serde reasons but I'll be able to get rid of that with hex-conservative 0.2. 0.31 also let me get rid of a bunch of stupid `Amount`/`u64` conversions because it finally made `TxOut::amount` correctly typed.

So in conclusion 0.30 was a mess that I was only able to navigate after practicing on multiple crates (rust-miniscript and rust-elements notably), but the 0.31 changes were all straightforward and discoverable improvements. So we are getting better.

**Comment by tcharding** [comment]
Yeah 0.30.0 hurt everyone and I think left a bad taste in everyones mouth as well. I definitely regret that but I also think we are getting better.

**Comment by apoelstra** [comment]
I kinda think we should backport the `pub use hex_conservative as hex;` re-export to 0.30.x.

**Comment by tcharding** [comment]
That's a good idea, I can do that.

**Comment by Kixunil** [comment]
> I don't know if you worked much on larger Rust-based Bitcoin projects

I did and it uncovered even more problems. Just because you can write the code doesn't mean the code is good.

> _stability to me is way more important than getting the API 100% right_.

False dichotomy. It's not about stability vs right API but time preference. You appear to have higher time preference and it is odd that you use Rust which is generally geared towards people with lower.

> **I think a library like this should have as main goal to have as many eyes on the code as possible**.

It doesn't help to have an internally super-secure library if the users are going to shoot themselves into the foot at the first opportunity.

> And right now it is achieving the exact opposite.

It gets worse before it gets better. And I'd bet most people never looked at less common modules such as block filters. And I doubt your claim is true anyway. Like literally every time I look at some code I didn't see before I see a bunch of issues with it so if there were many reviewers beforehand how did they not see those things?

> none of which I think are active users of the library in applications

You literally met me at a conference with the whole team we're using this with. :) Not to mention I started contributing mainly related to electrs to which I also contributed a lot.

My usage helped shape this library a lot. If it wasn't midnight here I'd pull a stat how many LOC we removed upgrading to the new taproot API. But I remember well that the code got much cleaner. Also saved our ass because it found some forgotten address network checks.

> we released 0.30 without any hex support

That's quite an exaggeration. We didn't support hex for arbitrary slices/arrays but all our types that previously had hex kept it.

**Comment by tcharding** [comment]
FWIW this release's primary task is, and has been, creating the `units` crate. Next release's focus is to create `primitives`. Once this is done downstream should be able to start using `primitives` types and it should (fingers crossed) become possible for different projects to be on different versions of `rust-bitcoin`.

Can we close this issue now @stevenroose?

**Comment by apoelstra** [comment]
I'd like to see a release or two without grief before closing this.

I know you want to see the issue tracker thin out @tcharding but I think we're making meaningful progress here and should keep this open as a lodestar.

**Comment by tcharding** [comment]
Sweet as, no stress whats so ever.
